networkx/algorithms/dag.py

```diff
@@ -414,20 +421,22 @@ def dag_longest_path(G):
     --------
     dag_longest_path_length
     """
-    dist = {}  # stores [node, distance] pair
-    for node in nx.topological_sort(G):
-        # pairs of dist,node for all incoming edges
-        pairs = [(dist[v][0] + 1, v) for v in G.pred[node]]
-        if pairs:
-            dist[node] = max(pairs)
-        else:
-            dist[node] = (0, node)
-    node, (length, _) = max(dist.items(), key=lambda x: x[1])
+    dist = {} # stores {v : (length, u)}
+    for v in nx.topological_sort(G):
+        us = [(dist[u][0] + data.get(weight, default_weight), u)
+            for u, data in G.pred[v].items()]
+        # Use the best predecessor if there is one and its distance is non-negative, otherwise terminate.
+        maxu = max(us) if us else (0, v)
```

Comment:
Yes, I tried that out. The issue then is that ties are broken in a way that depends on the order of the elements in the hash table. That caused one of the previous unit tests to fail, because it chose a different arbitrary path of the same length. I didn't use that approach to maintain backwards compatibility.